### PR TITLE
Add RFC for combinable simulation triggers with `initial()`.

### PR DESCRIPTION
**Pull request by wanda-phi**
[Rendered](https://github.com/wanda-phi/rfcs/blob/sim-initial/text/0067-sim-initial.md)

### CODE REVIEW

File: text/0067-sim-initial.md [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+
+An increase in language complexity.
+
+The `.tick()` part is an obscure edge case.
+
+
+## Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+This RFC proposes to leave current semantics unchanged and add `.initial()` as something that has to be explicitly requested. Arguably, this can be seen as the wrong default. An alternative would be to make the "initial" behavior default whenever `async for` is involved. However, this in turn can be a very wrong default for other cases, and the flexibility of combinable triggers means that choosing a good default would require horrifying heuristics.
```
Comment (whitequark) [inline text/0067-sim-initial.md:102]: Right. I think it's worth mentioning that we're discussing this because we coupled a performance improvement to the use of the `async for` syntax, which is why it's not viable to dedicate the `async for` syntax to processes.

File: text/0067-sim-initial.md [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+
+The Verilog `always @*` construct has the same problem, resulting in pain and horrible workarounds whenever Verilog has to be emitted, including Amaranth's Verilog backend.
+
+The SystemVerilog `always_comb` construct implicitly performs an equivalent of `.initial()`.
+
+## Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+- is `.tick().initial()` worth the complexity budget?
+- should we make writing the wrong implementation of combinational logic harder, somehow?
```
Comment (whitequark) [inline text/0067-sim-initial.md:115]: Should we call it `.init()`, as a parallel to the `init=` argument, and unlike the Verilog construct?

File: text/0067-sim-initial.md [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+The `.initial()` method is meant only for use with `async for` — passing the result of `.initial()` to plain `await` is an error.
+
+
+## Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Both combinable and initial trigger objects are amended with one new method:
+
+- `initial()` 
+  - Create a new trigger object of the same type by copying the current object and appending an initial trigger. This causes the resulting trigger object to fire once immediately after first starting iteration by an `async for` loop. Awaiting such a trigger with plain `await` is an error. The `.initial()` trigger doesn't contribute any components to the yielded value.
```
Comment (whitequark) [inline text/0067-sim-initial.md:87]: It is obvious what `.changed(...).initial()` would return. What about `.tick(...).initial()` though? Can it race with other processes at time 0?

**Comment by whitequark** [comment]
My thoughts on the core of the RFC:

- Anyone who writes `async for _ in ctx.changed():` should be prepared to deal with glitches and make sure the body of the loop is either pure or at least idempotent, so it is probably safe to have `ctx.changed()` imply `.initial()`. I don't think we guarantee, or ever intend to guarantee, that `changed()` triggers are glitchless--even for testbenches!
- The specific narrow case of `async for _ in ctx.tick():` entering the loop initially if an async reset is asserted at the beginning of the simulation feels so narrowly scoped as to seem completely harmless; perhaps even more so than the former case. I cannot imagine any case at all where it would pose a problem--indeed, it's indistinguishable from the case of "all signals initialize to 0 and then a buffer added using a process sets an async reset to 1".

File: text/0067-sim-initial.md (hunk 3 above)
Comment (wanda-phi) [reply text/0067-sim-initial.md:87]: I don't know what kind of races you mean; all races should be prevented by other processes' changes not happening until the next delta cycle, right?

Comment (whitequark) [reply text/0067-sim-initial.md:87]: It's a little more complicated than that. Normally, races (e.g. pulse on clk and rst simultaneously) are avoided because async signals don't toggle simultaneously with other signals. For example, clocks will be usually toggled by `add_clock` while everything else in the design is waiting for that tick, and resets will be usually asserted by testbenches as `ctx.set(rst, 1)` that doesn't change any other signal unless you opt into it via `ctx.set(Cat(...), ...)`.

However, things are weirder at time 0. Consider a situation where many combinational processes are used at time 0, all with `.initial()` or equivalent. These will all run in the same delta cycle, and may change a large amount of signals in that cycle. If several of those signals are a part of the same async control set you get a race.

In short, we don't actually *prevent* races, we make them harder to accidentally trigger via UI design, and in case of `.initial()` specifically the UI design part isn't there.

---

So knowing exactly what `.tick()` returns is important because the behavior during a race should be well-defined. Is it neither `clk` nor `rst? Is it `rst` but not `clk`? On careful reading I conclude that it's the latter but the RFC text should definitely say that explicitly.

**Comment by whitequark** [comment]
I have an alternate proposal that achieves the same underlying goal (well-defined behavior at time 0):

1. `ctx.<any>.changed(...).<any>` acts as-if there was a change to an unspecified signal if it is awaited (via `await` or `async for`) at time 0;
2. `ctx.tick(domain).<any>` acts as-if there was an assertion of the asynchronous reset if `domain` has an asynchronous reset, the initial value of this asynchronous reset is 1, and it is awaited (via `await` or `async for`) at time 0.

The case (2) is niche enough that I can't imagine any change to the surface language pulling its weight in terms of complexity. The case (1) is less unambiguous, but since we don't guarantee absence of glitches, avoiding a very narrow sliver of spurious wakeups doesn't seem worth it.

**Comment by wanda-phi** [comment]
abandoning in favor of an already-merged simpler fix (ie. [3d2cd15435bed62f026af64ba9c99b6f72759a97](https://github.com/amaranth-lang/amaranth/commit/3d2cd15435bed62f026af64ba9c99b6f72759a97))
